src/man-pages/man/man2/flock.2
.\" Copyright, the authors of the Linux man-pages project
.\"
.\" SPDX-License-Identifier: Linux-man-pages-copyleft
.\"
.\" FIXME Maybe document LOCK_MAND, LOCK_RW, LOCK_READ, LOCK_WRITE
.\" which only have effect for SAMBA.
.\"
.TH flock 2 2025-09-21 "Linux man-pages (unreleased)"
.SH NAME
flock \- apply or remove an advisory lock on an open file
.SH LIBRARY
Standard C library
.RI ( libc ,\~ \-lc )
.SH SYNOPSIS
.nf
.B #include <sys/file.h>
.P
.BI "int flock(int " fd ", int " op );
.fi
.SH DESCRIPTION
Apply or remove an advisory lock on the open file specified by
.IR fd .
The argument
.I op
is one of the following:
.RS 4
.TP 9
.B LOCK_SH
Place a shared lock.
More than one process may hold a shared lock for a given file
at a given time.
.TP
.B LOCK_EX
Place an exclusive lock.
Only one process may hold an exclusive lock for a given
file at a given time.
.TP
.B LOCK_UN
Remove an existing lock held by this process.
.RE
.P
A call to
.BR flock ()
may block if an incompatible lock is held by another process.
To make a nonblocking request, include
.B LOCK_NB
(by ORing)
with any of the above operations.
.P
A single file may not simultaneously have both shared and exclusive locks.
.P
Locks created by
.BR flock ()
are associated with an open file description (see
.BR open (2)).
This means that duplicate file descriptors (created by, for example,
.BR fork (2)
or
.BR dup (2))
refer to the same lock, and this lock may be modified
or released using any of these file descriptors.
Furthermore, the lock is released either by an explicit
.B LOCK_UN
operation on any of these duplicate file descriptors, or when all
such file descriptors have been closed.
.P
If a process uses
.BR open (2)
(or similar) to obtain more than one file descriptor for the same file,
these file descriptors are treated independently by
.BR flock ().
An attempt to lock the file using one of these file descriptors
may be denied by a lock that the calling process has
already placed via another file descriptor.
.P
A process may hold only one type of lock (shared or exclusive)
on a file.
Subsequent
.BR flock ()
calls on an already locked file will convert an existing lock to the new
lock mode.
.P
Locks created by
.BR flock ()
are preserved across an
.BR execve (2).
.P
A shared or exclusive lock can be placed on a file regardless of the
mode in which the file was opened.
.SH RETURN VALUE
On success, zero is returned.
On error, \-1 is returned, and
.I errno
is set to indicate the error.
.SH ERRORS
.TP
.B EBADF
.I fd
is not an open file descriptor.
.TP
.B EINTR
While waiting to acquire a lock, the call was interrupted by
delivery of a signal caught by a handler;
see
.BR signal (7).
.TP
.B EINVAL
.I op
is invalid.
.TP
.B ENOLCK
The kernel ran out of memory for allocating lock records.
.TP
.B EWOULDBLOCK
The file is locked and the
.B LOCK_NB
flag was selected.
.SH VERSIONS
Since Linux 2.0,
.BR flock ()
is implemented as a system call in its own right rather
than being emulated in the GNU C library as a call to
.BR fcntl (2).
With this implementation,
there is no interaction between the types of lock
placed by
.BR flock ()
and
.BR fcntl (2),
and
.BR flock ()
does not detect deadlock.
(Note, however, that on some systems, such as the modern BSDs,
.\" E.g., according to the flock(2) man page, FreeBSD since at least 5.3
.BR flock ()
and
.BR fcntl (2)
locks
.I do
interact with one another.)
.SS CIFS details
Up to Linux 5.4,
.BR flock ()
is not propagated over SMB.
A file with such locks will not appear locked for remote clients.
.P
Since Linux 5.5,
.BR flock ()
locks are emulated with SMB byte-range locks on the entire file.
Similarly to NFS, this means that
.BR fcntl (2)
and
.BR flock ()
locks interact with one another.
Another important side-effect is that the locks are not advisory anymore:
any IO on a locked file will always fail with
.B EACCES
when done from a separate file descriptor.
This difference originates from the design of locks in the SMB protocol,
which provides mandatory locking semantics.
.P
Remote and mandatory locking semantics may vary with
SMB protocol, mount options and server type.
See
.BR mount.cifs (8)
for additional information.
.SH STANDARDS
BSD.
.SH HISTORY
4.4BSD (the
.BR flock ()
call first appeared in 4.2BSD).
A version of
.BR flock (),
possibly implemented in terms of
.BR fcntl (2),
appears on most UNIX systems.
.SS NFS details
Up to Linux 2.6.11,
.BR flock ()
does not lock files over NFS
(i.e., the scope of locks was limited to the local system).
Instead, one could use
.BR fcntl (2)
byte-range locking, which does work over NFS,
given a sufficiently recent version of
Linux and a server which supports locking.
.P
Since Linux 2.6.12, NFS clients support
.BR flock ()
locks by emulating them as
.BR fcntl (2)
byte-range locks on the entire file.
This means that
.BR fcntl (2)
and
.BR flock ()
locks
.I do
interact with one another over NFS.
It also means that in order to place an exclusive lock,
the file must be opened for writing.
.P
Since Linux 2.6.37,
.\" commit 5eebde23223aeb0ad2d9e3be6590ff8bbfab0fc2
the kernel supports a compatibility mode that allows
.BR flock ()
locks (and also
.BR fcntl (2)
byte region locks) to be treated as local;
see the discussion of the
.I "local_lock"
option in
.BR nfs (5).
.SH NOTES
.BR flock ()
places advisory locks only;
given suitable permissions on a file,
a process is free to ignore the use of
.BR flock ()
and perform I/O on the file.
.P
.BR flock ()
and
.BR fcntl (2)
locks have different semantics with respect to forked processes and
.BR dup (2).
On systems that implement
.BR flock ()
using
.BR fcntl (2),
the semantics of
.BR flock ()
will be different from those described in this manual page.
.P
Converting a lock
(shared to exclusive, or vice versa) is not guaranteed to be atomic:
the existing lock is first removed, and then a new lock is established.
Between these two steps,
a pending lock request by another process may be granted,
with the result that the conversion either blocks, or fails if
.B LOCK_NB
was specified.
(This is the original BSD behavior,
and occurs on many other implementations.)
.\" Kernel 2.5.21 changed things a little: during lock conversion
.\" it is now the highest priority process that will get the lock -- mtk
.SH SEE ALSO
.BR flock (1),
.BR close (2),
.BR dup (2),
.BR execve (2),
.BR fcntl (2),
.BR fork (2),
.BR open (2),
.BR lockf (3),
.BR lslocks (8)
.P
.I Documentation/filesystems/locks.txt
in the Linux kernel source tree
.RI ( Documentation/locks.txt
in older kernels)


man-pages/Changes
==================== Changes in man-pages-6.16 ====================

Released: 2025-10-29, Aldaya


New and rewritten pages
-----------------------

man2/
	fsconfig.2
	fsmount.2
	fsopen.2
	fspick.2
	move_mount.2
	open_tree.2

man2const/
	PR_FUTEX_HASH_GET_SLOTS.2const
	PR_FUTEX_HASH_SET_SLOTS.2const

man2type/
	mount_attr.2type			(split from mount_setattr(2))

man3/
	_Fork.3
	countof.3
	memeq.3
	streq.3


Newly documented interfaces in existing pages
---------------------------------------------

man2/
	mount_setattr.2
		glibc wrapper

	pidfd_open.2
		PIDFD_THREAD

	pidfd_send_signal.2
		PIDFD_SIGNAL_THREAD
		PIDFD_SIGNAL_THREAD_GROUP
		PIDFD_SIGNAL_PROCESS_GROUP

	quotactl.2
		quotactl_fd(2)

	readv.2
		RWF_DONTCACHE

	sigaction.2
		SI_ASYNCNL
		FPE_FLTUNK
		FPE_CONDTRAP
		SEGV_ACCADI
		SEGV_ADIDERR
		SEGV_ADIPERR
		SEGV_MTEAERR
		SEGV_MTESERR
		SEGV_CPERR
		TRAP_UNK
		TRAP_PERF
		SYS_USER_DISPATCH
		EMT_TAGOVF

man3/
	abs.3
		umaxabs(3)

	opendir.3
		ENAMETOOLONG

man7/
	operator.7
		(type){}
		alignof
		_Countof

	rtnetlink.7
		RTAX_UNSPEC
		RTAX_LOCK
		RTAX_MTU
		RTAX_WINDOW
		RTAX_RTT
		RTAX_RTTVAR
		RTAX_SSTHRESH
		RTAX_CWND
		RTAX_ADVMSS
		RTAX_REORDERING
		RTAX_HOPLIMIT
		RTAX_INITCWND
		RTAX_FEATURES
		RTAX_RTO_MIN
		RTAX_INITRWND
		RTAX_QUICKACK
		RTAX_CC_ALGO
		RTAX_FASTOPEN_NO_COOKIE

	sched.7
		CONFIG_PREEMPT
		CONFIG_PREEMPT_LAZY

	tcp.7
		TCP_SAVE_SYN
		TCP_SAVED_SYN


New and changed links
---------------------

man2/
	open_tree_attr.2			(open_tree(2))
	quotactl_fd.2				(quotactl(2))
man3/
	umaxabs.3				(abs(3))


Global changes
--------------

-  man/
   -  Update documentation for POSIX.1-2024 (still more work to do).
   -  SYNOPSIS: Use GNU forward-declarations of parameters for sizes of
      array parameters.
   -  Improve style of source code for tbl(1) tables.
   -  man2/: Use the common name 'fd' instead of 'fildes'.
   -  Use semantic newlines.
   -  EXAMPLES
      -  Use err(3) and errc(3bsd) instead of similar macros.
      -  Fix includes (thanks to iwyu(1)).
      -  Use NITEMS() consistently.
      -  Don't name unused parameters.
-  share/mk/
   -  Add support for the chapter man3attr.
   -  help: Document the usual targets.
   -  lint-man-blank: Add target to lint about blank lines.
   -  lint-man-ws: Diagnose spurious use of white space.
   -  lint-man-quote: Diagnose an unmatched quote.
   -  lint-man-poems: Diagnose (lack of) semantic newlines.
   -  lint-man-dash: Diagnose unescaped dashes
   -  Merge handling of man(7) and mdoc(7) pages.
   -  lint, build, check: Move exceptions to separate files.
   -  Remove unused variables.
   -  Unify local variable names.
   -  Silence diagnostics about unnamed parameters.
   -  configure/build-depends/:
      -  Add awk/awk.
      -  Add pcre2-utils/pcre2grep.


Changes to individual files
---------------------------

-  src/bin/
   -  sortman(1): Fix order of intro(*) pages for subsections.
   -  diffman-git(1): Show sections as hunk context.

The manual pages and other files in the repository have been improved
beyond what this changelog covers.  To learn more about changes applied
to individual pages, or the authors of changes, use git(1).


src/mandb/lib/flock.c
/* Copyright (C) 1992 Free Software Foundation, Inc.
This file is part of the GNU C Library.

The GNU C Library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.

The GNU C Library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with the GNU C Library; see the file COPYING.LIB.  If
not, write to the Free Software Foundation, Inc., 675 Mass Ave,
Cambridge, MA 02139, USA.  */

/* This file implements the `flock' function in terms of the POSIX.1 `fcntl'
   locking mechanism.  In 4BSD, these are two incompatible locking mechanisms,
   perhaps with different semantics?  */

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#include "ansidecl.h"
#include "flock.h"
#include <errno.h>

#ifndef STDC_HEADERS
extern int errno;
#endif

#if HAVE_UNISTD_H
#  include <unistd.h>
#endif
#if HAVE_FCNTL_H
#  include <fcntl.h>
#endif
#if HAVE_SYS_FILE_H
#  include <sys/file.h>
#endif

/* Apply or remove an advisory lock, according to OPERATION,
   on the file FD refers to.  */
int
DEFUN(flock, (fd, operation), int fd AND int operation)
{
  struct flock lbuf;

  switch (operation & ~LOCK_NB)
    {
    case LOCK_SH:
      lbuf.l_type = F_RDLCK;
      break;
    case LOCK_EX:
      lbuf.l_type = F_WRLCK;
      break;
    case LOCK_UN:
      lbuf.l_type = F_UNLCK;
      break;
    default:
      errno = EINVAL;
      return -1;
    }

  lbuf.l_whence = SEEK_SET;
  lbuf.l_start = lbuf.l_len = 0L; /* Lock the whole file.  */

  return fcntl (fd, (operation & LOCK_NB) ? F_SETLK : F_SETLKW, &lbuf);
}
